server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java

```diff
@@ -0,0 +1,116 @@
+
+    public static void refreshReplicaShard(
+        WriteRequest.RefreshPolicy policy,
+        IndexShard indexShard,
+        Translog.Location location,
+        ActionListener<Boolean> listener
+    ) {
+        doRefreshShard(policy, false, null, indexShard, location, listener);
+    }
+
+    private static void doRefreshShard(
+        WriteRequest.RefreshPolicy policy,
+        boolean isPrimary,
+        @Nullable TransportService transportService,
+        IndexShard indexShard,
+        Translog.Location location,
+        ActionListener<Boolean> listener
+    ) {
+        switch (policy) {
+            case NONE -> listener.onResponse(false);
+            case WAIT_UNTIL -> {
+                if (location != null) {
+                    indexShard.addRefreshListener(
+                        location,
+                        refreshResult -> afterRefresh(
+                            indexShard,
+                            isPrimary,
+                            transportService,
+                            listener,
+                            refreshResult.refreshForced(),
+                            refreshResult.generation()
+                        )
+                    );
+                } else {
+                    listener.onResponse(false);
+                }
+            }
+            case IMMEDIATE -> {
+                Engine.RefreshResult refreshResult = indexShard.refresh("refresh_flag_index");
+                afterRefresh(indexShard, isPrimary, transportService, listener, true, refreshResult.generation());
+            }
+            default -> throw new IllegalArgumentException("unknown refresh policy: " + policy);
+        }
+    }
+
+    private static void afterRefresh(
+        IndexShard indexShard,
+        boolean isPrimary,
+        @Nullable TransportService transportService,
+        ActionListener<Boolean> listener,
+        boolean wasForced,
+        long generation
+    ) {
+        if (isPrimary && hasUnpromotables(indexShard)) {
+            assert transportService != null : "TransportService cannot be null if unpromotables present";
+
+            // Integrate #93600
+            // UnpromotableShardRefreshRequest unpromotableReplicaRequest = new UnpromotableShardRefreshRequest(
+            // indexShard.getReplicationGroup().getRoutingTable(),
+            // result.generation()
+            // );
+            // transportService.sendRequest(
+            // transportService.getLocalNode(),
+            // TransportUnpromotableShardRefreshAction.NAME,
+            // unpromotableReplicaRequest,
+            // new ActionListenerResponseHandler<>(
+            // listener.delegateFailure((l, r) -> l.onResponse(null)),
+            // (in) -> ActionResponse.Empty.INSTANCE,
+            // ThreadPool.Names.REFRESH
+            // )
+            // );
+            // TODO: Remove
+            listener.onResponse(wasForced);
+        } else {
+            listener.onResponse(wasForced);
```

Comment:
Using ActionListener.completeWith() seems that it could remove having two `onResponse`s.